zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -135,52 +137,48 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
+        uint end = start + length - 1;
 
         // The remaining bits could be written over two field elements
         // as such we compute the indices of the first and last field elements
         // they are located at
         uint index_start_bytes32 = 1 + 1 + 2*jsIn + jsOut + (offset+start)/field_capacity;
-        uint index_end_bytes32 = 1 + 1 + 2*jsIn + jsOut + (offset+end)/field_capacity;
+        uint index_end_bytes32 = index_start_bytes32 + (length-1)/field_capacity;
 
         // We compute the padding length in the field elements containing the remaining bits
-        // The padding is either digest_length - field_capacity (if we fill a whole field element)
+        // The padding is either length_extra_bits (if we fill a whole field element)
         // or digest_length - len(v_in || v_out || hsig || {sn} || {cm} || {h})
-        uint padding_start = digest_length - field_capacity;
-        uint padding_end = padding_start;
-        // We check if we index_end_bytes32 points to the last byte.
-        // The last byte may not be full, and thus have more padding.
-        if (primary_inputs.length-1 == index_end_bytes32) {
-            padding_end = digest_length - ( (offset + (digest_length - field_capacity)*(2*jsIn + jsOut)) % field_capacity);
-        }
-        // We check if the last byte is the same as the first byte, if so we update the padding.
-        if (index_end_bytes32 == index_start_bytes32) {
-            padding_start = padding_end;
-        }
+        uint potential_padding = digest_length - ((offset + length_extra_bits*(2*jsIn + jsOut)) % field_capacity);
+        uint padded_start = start + offset + ((primary_inputs.length-1 == index_start_bytes32) ? potential_padding : length_extra_bits);
+        uint padded_end = end + offset + ((primary_inputs.length-1 == index_end_bytes32) ? potential_padding : length_extra_bits);
 
         // We can now compute the bytes; indices of the remaining bits in both field elements
         // and retrieve them
-        uint index_start_byte = (padding_start+offset+start)/8;
-        uint8 start_byte = uint8(bytes32(primary_inputs[index_start_bytes32])[index_start_byte]);
-        uint index_end_byte = (padding_end+offset+end)/8;
-        uint8 end_byte = uint8(bytes32(primary_inputs[index_end_bytes32])[index_end_byte]);
+        uint8 start_byte = uint8(bytes32(primary_inputs[index_start_bytes32])[padded_start/8]);
+        uint8 end_byte = uint8(bytes32(primary_inputs[index_end_bytes32])[padded_end/8]);
 
         // We now recombine the two bytes
         // We multiply start_byte by 256 = 2**8.
-        res = start_byte*(2**8)+end_byte;
+        uint16 res = start_byte*(2**8) + end_byte;
         // And remove any unneeded bits
-        uint index_start_bit = (offset + padding_start + start) % 8;
-        res = res * uint16(2**index_start_bit);
-        res = res / uint16(2**(16 - (end - start + 1)));
+        res = res * uint16(2**(padded_start % 8));
```

Comment:
` uint padded_start = start + offset + ((primary_inputs.length-1 == index_start_bytes32) ? potential_padding : length_extra_bits);`

and now you do:
> res = res * uint16(2**(padded_start % 8));

If this supposed to be equivalent to what you did in your previous commit:
```
uint index_start_bit = (offset + padding_start + start) % 8;
```
Because it does not seem to be.

Reply:
It should be equivalent:
We have that
`padding_start = digest_length - ( (offset + (digest_length - field_capacity)*(2*jsIn + jsOut)) % field_capacity`  if `primary_inputs.length-1 == index_start_bytes32`
else 
`uint potential_padding = digest_length - ((offset + length_extra_bits*(2*jsIn + jsOut)) % field_capacity);` is the same as before and by definition `length_extra_bits` equals `digest_length - field_capacity`

The main difference is than now if we ask for a length of 0 we directly leave.

On the other side we have:
`uint padded_start = start + offset + ((primary_inputs.length-1 == index_start_bytes32) ? potential_padding : length_extra_bits);`
So "padding start" should be: `((primary_inputs.length-1 == index_start_bytes32) ? potential_padding : length_extra_bits);`
`potential_padding`